### PR TITLE
Migrate to null safety

### PR DESCRIPTION
**Pull request by kevmoo**

### CODE REVIEW

**Comment by natebosch** [comment]
Are we ready for this to go into `master` branch?

File: example/hash.dart [hunk 1]
```diff
@@ -10,12 +10,13 @@ import 'package:crypto/crypto.dart';
 final _usage = 'Usage: dart hash.dart <md5|sha1|sha256> <input_filename>';
 
 Future main(List<String> args) async {
-  if (args == null || args.length != 2) {
+  if (args.length != 2) {
     print(_usage);
-    exit(1);
+    exitCode = 64;
```
[the remark this answers is not in the document]
Comment (kevmoo) [reply example/hash.dart:15]: See my other PR.

Agree, exit should have a return of `Never` – will file!

[the remark this answers is not in the document]
Comment (kevmoo) [reply example/hash.dart:15]: https://github.com/dart-lang/crypto/pull/91

[the remark this answers is not in the document]
Comment (kevmoo) [reply example/hash.dart:15]: And https://github.com/dart-lang/sdk/issues/42006

**Comment by kevmoo** [comment]
> Are we ready for this to go into `master` branch?

Let's talk about the overall plan here...

**Comment by kevmoo** [comment]
@natebosch PTAL

File: lib/src/digest_sink.dart [hunk 2]
```diff
@@ -7,24 +7,18 @@ import 'digest.dart';
...
   void add(Digest value) {
-    assert(_value == null);
     _value = value;
   }
 
   @override
-  void close() {
-    assert(_value != null);
-  }
+  void close() {}
```
Comment (natebosch) [inline lib/src/digest_sink.dart:23]: I think this could be removed instead

Comment (natebosch) [reply lib/src/digest_sink.dart:23]: cc @jakemac53 - are you concerned at that we're losing this assert, or that we're turning debug only failures into always failures? I think it's probably OK but I'm curious what you think.

Comment (kevmoo) [reply lib/src/digest_sink.dart:23]: seems like it IS needed – it's abstract.

Comment (natebosch) [reply lib/src/digest_sink.dart:23]: We could consider a read of the variable to regain the check that it was assigned at some point. We could even do it in an `assert` if we wanted. I don't think we can replace the other one though.

```dart
void close() {
  assert(() {
    _value;
    return true;
  });
}
```

@leafpetersen - does using a `late final` field as a statement have the effect of ensuring it was assigned?

File: example/example.dart [hunk 3]
```diff
@@ -16,7 +16,7 @@ Future main(List<String> args) async {
     return;
   }
 
-  late Hash hasher;
+  Hash hasher;
```
Comment (kevmoo) [inline example/example.dart]: yay for definite assignment!

File: lib/src/digest_sink.dart [hunk 4]
```diff
@@ -9,7 +9,7 @@ class DigestSink extends Sink<Digest> {
   /// The value added to the sink, if any.
   Digest get value => _value;
 
-  late Digest _value;
+  late final Digest _value;
```
Comment (kevmoo) [inline lib/src/digest_sink.dart:12]: get us back to "assign once"

File: lib/src/digest_sink.dart (hunk 2 above)
Comment (leafpetersen) [reply lib/src/digest_sink.dart:23]: yes.

**Pull request merged**